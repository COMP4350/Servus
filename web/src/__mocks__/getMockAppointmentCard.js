function getMockAppointmentCard() {
    return {
        buyer: 'Risto',
        provider: 'Arveend',
        service_id: '1001',
        created_at: new Date(100),
        booked_time: new Date(500),
    }
}

export const mockAppointmentCard = getMockAppointmentCard();