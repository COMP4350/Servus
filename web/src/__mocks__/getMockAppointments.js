function getMockAppointments() {
    return [
        {
            buyer: 'Risto',
            provider: 'Arveend',
            service_id: '1001',
            created_at: new Date(100),
            booked_time: new Date(500),
        },
        {
            buyer: 'Arveend',
            provider: 'Risto',
            service_id: '1002',
            created_at: new Date(700),
            booked_time: new Date(1200),
        },
        {
            buyer: 'Caden',
            provider: 'Andy',
            service_id: '1003',
            created_at: new Date(1400),
            booked_time: new Date(1900),
        },
    ]
}

function getMockAppointmentsCookie() {
    return {
        username: 'testusername',
    }
}

export const mockAppointments = getMockAppointments();
export const mockAppointmentsCookie = getMockAppointmentsCookie();