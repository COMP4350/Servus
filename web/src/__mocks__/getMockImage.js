function getMockImage() {
    return {
        ownerUsername: 'riki',
        uploadDate: new Date(2012, 1, 1),
        imageName: 'example_image.jpeg',
        imageData: 'the/link/to/data',
        profilePicture: false,
    };
}

export const mockImage = getMockImage();
