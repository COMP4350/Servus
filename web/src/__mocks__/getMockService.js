function getMockService() {
    return {
        _id: '1001',
        name: "Risto's Painting",
        provider: 'Risto',
        description: 'Risto paints you',
        cost: 100,
        duration: '0100',
        availability: [{ weekday: 1, start_time: '0800', end_time: '1800' }],
        ratings: [{ username: 'arvind', rating: 5 }],
    };
}

export const mockService = getMockService();
