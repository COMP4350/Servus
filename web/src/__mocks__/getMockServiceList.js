function getMockServiceList() {
    return [
        {
            name: "Risto's Painting",
            provider: 'Risto',
            description: 'Risto paints you',
            cost: 100,
            duration: 60,
        },
        {
            name: "Reesto's Painting",
            provider: 'Reesto',
            description: 'Reesto paints yee',
            cost: 120,
            duration: 80,
        },
        {
            name: "Rasta's Painting",
            provider: 'Rasta',
            description: 'Rasta paints yaa',
            cost: 150,
            duration: 90,
        },
    ];
}

export const mockServiceList = getMockServiceList();
