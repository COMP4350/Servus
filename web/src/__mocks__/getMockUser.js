function getMockUser() {
    return {
        username: 'testuser',
        firstName: 'test',
        lastName: 'user',
        password: '123',
        bio: 'im a test user',
    };
}

export const mockUser = getMockUser();
