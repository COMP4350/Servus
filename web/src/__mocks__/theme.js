import { createMuiTheme } from '@material-ui/core/';

export default createMuiTheme({
    palette: {
        primary: {
            main: '#647AA3',
        },
        secondary: {
            main: '#EC5732',
        },
        default: {
            main: '#FFF2EB',
        },
    },
    background: {
        dark: '#151515',
        main: '#272727',
    },
    typographs: {
        fontFamily: ['Roboto'],
    },
});
