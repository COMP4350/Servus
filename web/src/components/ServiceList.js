import React, { useState, useEffect } from 'react';
import axios from 'axios';
import {
    makeStyles,
    InputBase,
    Divider,
    IconButton,
    List,
    ListItem,
    Paper,
    Chip,
    Typography,
} from '@material-ui/core';
import { FilterList, Search } from '@material-ui/icons/';
import tagNames from './FilterList';
import ServiceCard from './ServiceCard';
import useForm from '../hooks/useForm';

// Styles for the service list (left side) panel.
const useStyles = makeStyles(theme => ({
    rootPanel: {
        [theme.breakpoints.up('xs')]: {
            minWidth: '384px',
            width: '384px',
            height: '100%',
        },
        'background-color': theme.background.dark,
        'overflow-y': 'scroll',
        [theme.breakpoints.down('xs')]: {
            width: '100%',
            height: '100%',
        },
    },
    rootList: {
        padding: '8px',
        width: 'auto',
        height: 'auto',
        'overflow-x': 'hidden',
    },
    filters: {
        'background-color': theme.background.dark,
        margin: '4%',
        width: '92%',
    },
    searchBar: {
        margin: '4%',
        padding: '0',
        width: '92%',
        height: '40px',
        'flex-wrap': 'nowrap',
        'background-color': theme.background.main,
        'border-radius': '12px',
    },
    searchIcon: {
        cursor: 'pointer',
        color: 'white',
        width: '12.5%',
        padding: 10,
    },
    searchInput: {
        width: '75%',
        flex: 1,
        color: 'white',
    },
    tagChip: {
        margin: '2px',
    },
    divider: {
        height: 28,
        margin: 4,
        color: 'white',
    },
    bullet: {
        'padding-left': '5px',
        'padding-right': '5px',
        color: '#ec5732',
    },
}));

// Styles for the list item contents.
const listItemClass = makeStyles(theme => ({
    root: {
        '& p': {
            color: '#545454',
        },
        padding: '0',
        'padding-top': '4px',
        'padding-bottom': '4px',
        height: 'auto',
        width: 'auto',
    },
    selected: {
        '& p': {
            color: 'white',
        },
    },
    divider: {
        'border-color': theme.background.main,
    },
}));

// Service List component.
const ServiceList = props => {
    const [services, setServices] = useState(null);
    const [selectedIndex, setSelectedIndex] = useState(-1);
    const [serviceTags, setServiceTags] = useState(false);
    const [activeFilters] = useState([]);
    const [change, setChange] = useState(false);
    const [searchForm, setSearchForm] = useForm({ search: '' });
    const [errorText, setErrorText] = useState('');

    const classes = useStyles();
    const style = listItemClass();
    const getServices = async () => {
        try {
            const response = await axios.post(`/services/filter`, {
                tags: activeFilters,
                search: searchForm.search,
            });
            setServices(response.data.result);
        } catch {
            err => setErrorText(err);
        }
    };

    const handleListItemClick = (e, i) => {
        setSelectedIndex(i);
    };

    const openFilterList = () => {
        setServiceTags(!serviceTags);
    };

    // Show the chips used to set tag filters.
    const loadChips = () => {
        return (
            <Paper className={classes.filters}>
                {serviceTags
                    ? tagNames.map((tag, i) => {
                          return (
                              <Chip
                                  size="small"
                                  key={i}
                                  label={tag}
                                  onClick={() => addFilter(tag)}
                                  className={classes.tagChip}
                                  color={
                                      activeFilters.includes(tag)
                                          ? 'primary'
                                          : 'default'
                                  }
                              />
                          );
                      })
                    : null}
            </Paper>
        );
    };

    // Add a tag to the current list of tag filters.
    const addFilter = tag => {
        if (activeFilters.includes(tag)) {
            const index = activeFilters.indexOf(tag);
            if (index > -1) activeFilters.splice(index, 1);
        } else {
            activeFilters.push(tag);
        }
        setChange(!change);
        setSelectedIndex(-1);
    };

    const search = () => {
        console.log(`searching: ${searchForm.search}`);
    };

    const onServiceFormChange = e => {
        setSearchForm(e);
    };

    useEffect(() => {
        getServices();
        loadChips();
    }, [serviceTags, change, searchForm.search]);

    return (
        <Paper className={classes.rootPanel}>
            <Paper className={classes.searchBar}>
                <IconButton
                    className={classes.searchIcon}
                    onClick={openFilterList}
                    aria-label="menu">
                    <FilterList />
                </IconButton>
                <InputBase
                    className={classes.searchInput}
                    placeholder="Search"
                    name="search"
                    onChange={onServiceFormChange}
                    value={searchForm.search}
                    onKeyDown={e =>
                        e.key === 'Enter' ? search() : console.log(e.key)
                    }
                />
                <IconButton
                    type="submit"
                    className={classes.searchIcon}
                    onClick={search}
                    aria-label="search">
                    <Search />
                </IconButton>
                <Divider className={classes.divider} orientation="vertical" />
            </Paper>
            {loadChips()}
            <List className={classes.rootList}>
                {services
                    ? services.map((service, index) => {
                          if (selectedIndex == index) {
                              props.setSelectedService(service);
                          }
                          return (
                          <div
                            data-cy={`service_${index}`}
                            key={index}>
                              <ListItem 
                                  classes={style}
                                  onClick={e => handleListItemClick(e, index)}
                                  selected={selectedIndex == index}
                                  divider={true}
                                  >
                                  <ServiceCard
                                      service={service}
                                      index={index}
                                      className={classes.serviceCard}
                                      selected={selectedIndex == index}
                                      expand={false}
                                  />
                              </ListItem>
                          </div>
                          );
                      })
                    : <Typography>{errorText}</Typography>}
            </List>
        </Paper>
    );
};

export default ServiceList;
