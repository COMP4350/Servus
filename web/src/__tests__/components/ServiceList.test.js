import React from 'react';
import { act, render } from '@testing-library/react';
import ServiceList from '../../components/ServiceList';
import { mockServiceList } from '../../__mocks__/getMockServiceList';
import mockAxios from 'axios';


describe('ServiceList', () => {
    it('renders correctly', async () => {
        const promise = Promise.resolve();
        let handleServiceUpdate = jest.fn(() => promise);

        mockAxios.get.mockImplementationOnce(() => Promise.resolve({
            data: { results: mockServiceList }
        }));

        const { asFragment } = render(<ServiceList setServices={handleServiceUpdate}/>);
        expect(asFragment()).toMatchSnapshot();
        expect(mockAxios.get).toHaveBeenCalledWith("/services/");
        expect(mockAxios.get).toHaveBeenCalledTimes(1);
        
        await act(() => promise);
    });
});
