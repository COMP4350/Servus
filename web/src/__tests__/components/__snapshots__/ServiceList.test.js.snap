// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceList renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root makeStyles-rootPanel-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiPaper-root makeStyles-searchBar-4 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <button
        aria-label="menu"
        class="MuiButtonBase-root MuiIconButton-root makeStyles-searchIcon-5"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div
        class="MuiInputBase-root makeStyles-searchInput-6"
      >
        <input
          class="MuiInputBase-input"
          name="search"
          placeholder="Search"
          type="text"
          value=""
        />
      </div>
      <button
        aria-label="search"
        class="MuiButtonBase-root MuiIconButton-root makeStyles-searchIcon-5"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <hr
        class="MuiDivider-root makeStyles-divider-8 MuiDivider-vertical"
      />
    </div>
    <div
      class="MuiPaper-root makeStyles-filters-3 MuiPaper-elevation1 MuiPaper-rounded"
    />
    <ul
      class="MuiList-root makeStyles-rootList-2 MuiList-padding"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
      />
    </ul>
  </div>
</DocumentFragment>
`;
